### DISCUSSION TITLE
Custom Data Binding Converter: Static resource not found

### DISCUSSION THREAD

**Discussion by reelthymeos** [AvaloniaUI/Avalonia#16696] [Q&A] [answered]
Hi, 

I have a UserControl that uses a Grid. I needed a method that would dynamically allow me to hide a given row on the data grid at runtime. I found a convenient solution[ here ](https://stackoverflow.com/a/21884516/25330516)on stack overflow that uses a Data Binding convertor. The example is based off classic WPF but I know that the Avalonia version is essentially the same thing. I created a class for this convertor, see below:

```
using System;
using System.Collections.Generic;
//using System.Windows.Data;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Avalonia.Controls;
using Avalonia.Data.Converters;

namespace myApp.Helper_Classes
{
	//[System.Windows.Data.ValueConversion(typeof(bool), typeof(GridLength))]
	public class BoolToGridRowHeightConverter : IValueConverter
	{
		public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
		{
			return ((bool)value == true) ? new GridLength(1, GridUnitType.Star) : new GridLength(0);
		}

		public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
		{    // Don't need any convert back
			return null;
		}
	}
}
```

I commented out the Attribute header as that was in the WPF version.  Not sure if that's needed. Then I wrote the following line for a row definition in my XAML file:

`<RowDefinition Height="{Binding Row1_Visible, Converter={StaticResource BoolToGridRowHeightConverter}}"></RowDefinition>`

The error I'm getting is "Static Resource BoolToGridRowHeightConverter" not found. 

I found a similiar question on this forum here: https://github.com/AvaloniaUI/Avalonia/discussions/15073 

It seems that I need to add some kind of reference to this class in the XAML, but I'm not sure what to add. I tried to add the following, mimicking the example in the docs:

```
<UserControl.Resources>
		<local:BoolToGridRowHeightConverter x:Key="BoolToGridRowHeightConverter" />
</UserControl.Resources>
```
But intellisense is telling me it doesn't know what "local" is referring to. How can I properly reference my convertor class? It is located in a namespace called "Helper_Classes"

Thanks!

**Comment by stevemonaco** [comment, marked as answer]
I typically have a namespace specifically for converters: `namespace myApp.Converters;`

`local:` here is an XML Namespace. It's something you need to define. Think of it as a `using` in C# to import a namespace, except you still need to explicitly specify it everywhere it's used. I'll rename it to `con` for converters instead.

```xml
<UserControl ...
    xmlns:con="using:myApp.Converters">
    <UserControl.Resources>
        <con:BoolToGridRowHeightConverter x:Key="BoolToGridRowHeightConverter" />
    </UserControl.Resources>
```

Upon reading more, you're mixing up `DataGrid` and `Grid`. These are two different controls with very different purposes. `RowDefinition` doesn't exist on `DataGrid`.

**Reply by reelthymeos** [reply]
Thanks @stevemonaco, I edited my original question, I meant Grid and not DataGrid.

**Comment by reelthymeos** [comment]
I implemented this for all the rows (I have 12 total rows). As a test case, I set the booleans = TRUE for rows 0-4, and false for the remaining rows. But when I launch the app, the grid still shows those remaining rows as small rows at the bottom. even though the converter sets the height to zero per the function. Any guesses why this might be happening?

![image](https://github.com/user-attachments/assets/a797e91c-983e-4978-8e48-c09788ffad95)

**Reply by IanRawley** [reply]
Likely 0 is less than whatever the default minimum height is, so each row ends up minimum height and draws what it can.  Try setting the RowDefinition MinHeight to 0 as well and see what happens.

**Reply by stevemonaco** [reply]
Not really sure. The `MinHeight` from the expander chevrons in the drop down box shouldn't be able to override the `Grid` row height.

I would check with DevTools (F12) at runtime. Check the layout pane and see why the control still has size.

As an aside, my typical approach is to instead assign these rows as `auto` height dimensions then bind `IsVisible` on the topmost control within each cell. This will collapse to 0 since they aren't rendered and have no size. Then again, I typically don't use `Grid` when dynamic layout like this is required.

**Reply by IanRawley** [reply]
Do you have the full definitions for the Grid layouts?  Row and Column definitions + SharedSizeGroup stuff?  A lot of size calculations are effected by those other things, and looking in DefinitionBase there're certainly parts that could override the MinHeight especially if SharedSizeGroups are involved.

But also yes, what SteveMonaco said.  Rather than getting dug into the Row/Column Definitions to try and hide things it's easier to just make the root of each cell / row / column hide its contents and let the Grid handle it with Row / Column size set to Auto.

**Reply by reelthymeos** [reply]
@stevemonaco @IanRawley I was originally going to go the route of binding my boolean to the IsVisible property of the controls for each cell and have the row size be auto... but simply doing it once at the row level (via the convertor) just seemed more straightforward (less code). But in any case, the IsVisible method is working for me now. Thanks for the help.

**Reply by stevemonaco** [reply]
It's possible to apply bindings via Style Selector. You would likely target via direct children, direct children of a specific control type, or by style class. You may need to put `x:DataType` on the `Style` depending on the scope you place it in.
